### PR TITLE
Add "Assign file type" checkbox to Save As dialog

### LINKED ISSUE #9515: Move "Set Save dialog file extension filter to *.*" pref. setting to the SaveAs dialog
### Description of the Issue

Pursuant to the direction [HERE](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9328#issuecomment-757176413) and [HERE](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9328#issuecomment-761301587), let's move this Preference setting:

image

to be on the _SaveAs_ dialog itself, appearing in a similar manner as the _Save Folder as Workspace_ checkbox when one does a _File_ > _Save Session..._ :

image

Exact text for this that should be shown for this on the _SaveAs_ dialog is TBD.  The existing text from the Preferences area is perhaps a bit too long to use.

### Debug Information
Notepad++ v7.9.3   (32-bit)
Build time : Feb 14 2021 - 03:12:31
Path : C:\.........\npp.portable\notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 1909
OS Build : 18363.1316
Current ANSI codepage : 1252

### PR DESCRIPTION
**Pull request by mere-human**
Fix #9515

### CODE REVIEW

**Comment by mere-human** [comment]
**How it works:**
_Save As_ dialog now has _Assign file type_ checkbox (default is ON).
If it's ON, the file type (in dialog drop-down) as well as file extension is set based on document language.
If checkbox is OFF, then the file type is set to `*.*` and the file extension isn't added.
The checkbox value is remembered between runs.

**Demo:**

https://user-images.githubusercontent.com/9664141/113690787-72e86c80-96d4-11eb-9031-1b7f23212306.mp4

**Comment by donho** [comment]
Just a suggestion:
Using "Apply current doc. language filter" (instead of "Assign file type") may more relevant for users.

**Comment by mere-human** [comment]
> Using "Apply current doc. language filter" (instead of "Assign file type") may be more relevant for users.

This seems too long as for me. At least, it could be "Apply file language filter" or something like this.

But there are more ideas. I'm not sure what is the best option yet.
Please, see the latest comment here https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9515#issuecomment-814378151 and links to community forum.

**Comment by mere-human** [comment]
I should probably remove the checkbox from the preferences.

File: PowerEditor/src/WinControls/Preference/preference.rc [hunk 1]
```diff
@@ -416,9 +416,8 @@ BEGIN
     CONTROL         "Peek on document map",IDC_CHECK_ENABLEDOCPEEKONMAP,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,269,71,140,10
 
 	// "Enable Notepad++ auto-updater" should be always the 1st item, because it'll be hidden if GUP.exe is absent
-    CONTROL         "Mute all sounds", IDC_CHECK_MUTE_SOUNDS, "Button", BS_AUTOCHECKBOX | BS_MULTILINE | WS_TABSTOP, 37, 94, 217, 10
+    CONTROL         "Mute all sounds", IDC_CHECK_MUTE_SOUNDS, "Button", BS_AUTOCHECKBOX | BS_MULTILINE | WS_TABSTOP, 37, 109, 217, 10
     CONTROL         "Enable Notepad++ auto-updater",IDC_CHECK_AUTOUPDATE,"Button",BS_AUTOCHECKBOX | BS_MULTILINE | WS_TABSTOP, 37, 78, 210, 10
```
Comment (donho) [inline PowerEditor/src/WinControls/Preference/preference.rc:420]: image

IDC_CHECK_AUTOUPDATE should be made lower too.

**Review by donho — COMMENTED** [review]
1. Remove getNppGUIMutable(). Use getNppGUI instead.
2. Change IDC_CHECK_AUTOUPDATE control's Y position.

**Review by donho — CHANGES_REQUESTED** [review]
1. Remove getNppGUIMutable(). Use getNppGUI instead.
2. Change IDC_CHECK_AUTOUPDATE control's Y position.

File: PowerEditor/src/Parameters.h [hunk 2]
```diff
@@ -1339,6 +1339,8 @@ class NppParameters final
 		return _nppGUI;
 	}
 
+	NppGUI& getNppGUIMutable() { return _nppGUI; }
```
[the remark this answers is not in the document]
Comment (mere-human) [reply PowerEditor/src/Parameters.h:1342]: `const_cast` is bad: https://stackoverflow.com/questions/10498292/c-why-is-const-cast-evil
What if I declare it as this?
`NppGUI& getNppGUI() { return _nppGUI; }`
Just a non-const overload. But the code wouldn't require `const_cast`.

[the remark this answers is not in the document]
Comment (donho) [reply PowerEditor/src/Parameters.h:1342]: @mere-human 
I see your point.
Please use `const NppGUI & getNppGUI() const;` as it is in the PR.
I will remove keyword `const` myself afterward.

[the remark this answers is not in the document]
Comment (mere-human) [reply PowerEditor/src/Parameters.h]: As you wish.

File: PowerEditor/src/WinControls/Preference/preference.rc [hunk 3]
```diff
@@ -416,9 +416,8 @@ BEGIN
     CONTROL         "Peek on document map",IDC_CHECK_ENABLEDOCPEEKONMAP,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,269,71,140,10
 
 	// "Enable Notepad++ auto-updater" should be always the 1st item, because it'll be hidden if GUP.exe is absent
-    CONTROL         "Mute all sounds", IDC_CHECK_MUTE_SOUNDS, "Button", BS_AUTOCHECKBOX | BS_MULTILINE | WS_TABSTOP, 37, 94, 217, 10
-    CONTROL         "Enable Notepad++ auto-updater",IDC_CHECK_AUTOUPDATE,"Button",BS_AUTOCHECKBOX | BS_MULTILINE | WS_TABSTOP, 37, 78, 210, 10
-    CONTROL         "Set Save dialog file extension filter to *.*",IDC_CHECK_SAVEDLGEXTFILTALLTYPES,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,37,109,267,10
+    CONTROL         "Enable Notepad++ auto-updater",IDC_CHECK_AUTOUPDATE,"Button",BS_AUTOCHECKBOX | BS_MULTILINE | WS_TABSTOP, 37, 94, 210, 10
```
Comment (mere-human) [inline PowerEditor/src/WinControls/Preference/preference.rc:419]: I moved it upper in the code so it corresponds to the comment that says "Enable Notepad++ auto-updater" should be always the 1st item".
It was me who added "Mute all sounds" in the wrong place 😬

**Review by donho — APPROVED** [review]

**Comment by sasumner** [comment]
@mere-human Hooray!  I like it.
I should cautiously add: "We'll see how it goes with users".

**Comment by mere-human** [comment]
> I should cautiously add: "We'll see how it goes with users".

Users don't stop to surprise me with what they request and how they're able break our features 😄

**Comment by sasumner** [comment]
>Users don't stop to surprise me with what they request and how they're able break our features

Yes, but this one is a very special case.  It seems to be one where we can't predict what will arise.
Hopefully we (you!) are "honing in" on the final solution (the one that generates no more complaints).
It has been a long saga (6+ years) starting with old/new style user questions/problems...and probably hundreds of related Community postings...sigh...

**Comment by sasumner** [comment]
@mere-human 

Can you update the user manual for this feature?
Having said that, I'm not sure the best area it belongs in.
On the upside, I think I was remiss and never added the setting you just removed from the _Preferences_, so nothing will need to be removed there.

**Comment by mere-human** [comment]
> @mere-human
> 
> Can you update the user manual for this feature?

Sure, I will check it. Should I write something like in my 2nd comment in this PR?

**Comment by sasumner** [comment]
>like in my 2nd comment in this PR?

Very reasonable.
I'll let you create your own issue in the user manual issue tracker.
Thanks!

**Comment by mere-human** [comment]
@pryrt Do you have an idea whether this should be documented and what section it belongs to?

**Comment by sasumner** [comment]
>whether this should be documented

I thought we established this.
A strong YES.  :-)

>what section it belongs to?

It may need a new section.
And yes @pryrt 's thoughts on that would be super to have.
Normally there is no need to discuss open/save dialogs as everyone that uses Windows knows how they work.  !!

**Comment by pryrt** [comment]
@mere-human / @sasumner ,

Have Alan Kilborn and I misunderstood the intention of the "Append extension" checkbox, as Alan describes [in the Forum here](https://community.notepad-plus-plus.org/post/66533)?  Because as I understood its intention, if you type a filename with no extension when the "Append extension" is checked, Alan and I expect it to automatically add the first extension from the current Save As type.  But in Alan's experiment on 8.0-RC2, it does _not_ append the extension if there is no extension in the file-name box.

**Comment by mere-human** [comment]
> But in Alan's experiment on 8.0-RC2, it does _not_ append the extension if there is no extension in the file-name box.

Indeed, it should work as described. But it shouldn't because there is a bug.
It seems like the checkbox works as intended only when the save dialog is shown for the first time since app launch.
All subsequent calls don't work.
I'm going to fix that.

**Comment by mere-human** [comment]
@pryrt  Does it work if you press Enter key or Alt+S?

**Comment by sasumner** [comment]
@mere-human 

I tried it:

Enter key:  works!  (.txt extension added automagically)
Alt+s:  doesn't work (no extension added)

**Comment by donho** [comment]
@mere-human So is it a regression due to this PR?
Any opened issue about it?

**Comment by mere-human** [comment]
> @mere-human So is it a regression due to this PR?
> Any opened issue about it?

Yes, it is a regression.
I haven't seen any issue about it. I'll create one to document everything related.

**Comment by sasumner** [comment]
>regression

Nitpick: I would call it a bug because this checkbox is new functionality -- nothing to "regress".

**Pull request closed without merging**